nvim/nvim.go

```diff
@@ -187,6 +188,30 @@ func ChildProcessLogf(logf func(string, ...interface{})) ChildProcessOption {
 	}}
 }
 
+// ChildProcessDisableEmbed disables the --embed flag of nvim.
+// This is needed for applications to use the RPC API.
+// See: https://neovim.io/doc/user/starting.html#--embed for details.
+func ChildProcessDisableEmbed() ChildProcessOption {
+	return ChildProcessOption{func(cpos *childProcessOptions) {
+		cpos.disableEmbed = true
+	}}
+}
+
+// appendEmbedFlagIfNeeded appends the --embed flag, if it is not yet added.
+// This behavior can be overriden by setting the ChildProcessDisableEmbed() process option.
+func appendEmbedFlagIfNeeded(cpos *childProcessOptions) {
+	for _, arg := range cpos.args {
+		if arg == "--embed" {
+			return
+		}
+	}
+	if !cpos.disableEmbed {
+		cpos.logf("[go-client/nvim] Warning: '--embed' flag missing, appending by default. It enables RPC calls via stdin/stdout. To disable this behavior, add ChildProcessDisableEmbed(). More info: https://neovim.io/doc/user/starting.html#--embed")
```

Comment:
Hmm. I guess we can't make this a fatal error, because that would be backwards-incompatible.

Reply:
Actually, adding `--embed` is also backwards-incompatible. So maybe this _should_ be a fatal error.

Reply:
Hey @justinmk this is a good discussion to have. To arrive at this approach, I had to make some assumptions on how often `--embed` is being used. Sharing my analysis here:

| Method | Description | Backward-incompatible? i.e. does it break applications that don't want `--embed`? | Comments |
|--------|--------|--------|--------|
| **Method 1:**<br/>Log Fatal | Hard exit, then recommend user to include `--embed` in their `NewChildProcess()` options | Yes | Tells user to fix their params to include `--embed`. If `--embed` is really not needed, they can override this panic by adding `ChildProcessDisableEmbed()` |
| **Method 2:**<br/>Log Warn | Soft warning, without setting `--embed` | No | This just warns but doesn't append a default. This is a safe approach, but also (imo) the least helpful.  |
| **Method 3:**<br/>Log Warn and set Default | Soft warning, append a sensible default `--embed` | Yes | Assumes that `--embed` is the likely behavior that user wants, so adds it by default. If `--embed` is really not needed, user can disable it with `ChildProcessDisableEmbed()`. | 

I first eliminated **Method 2** because a soft warning, although backward-compatible, is probably not what user wants. I assumed that `--embed` is needed in most setups. And so even if users upgrade to this version, it doesn't fix anything except produce a warning, which isn't as useful as the other 2 methods.

So **Method 1** and **Method 3** attempts to make it useful by assuming `--embed` is a **sensible default** for most cases. Both methods will be backwards-incompatible because it assumes `--embed` as the preferred behavior. Both methods give a flag `ChildProcessDisableEmbed()` to opt out from `--embed` if this assumption is false.

The difference is:

- **Method 1** panics and educates users to fix their `NewChildProcess()` to include the `--embed` args in their setup.

- **Method 3** warns and sets the default to include `--embed`. 

Good thing about **Method 1** is it's explicit, however, it continues to make the `go-client` harder to setup/use because users will need to know that `--embed` is needed to make RPC calls. But as we saw from the related Issue #104 , it's actually not obvious that `--embed` flag needs to be passed.

I finally went with **Method 3** as it can simplify the usage of `go-client` by setting `--embed` by default. This means users don't need to know about the existence of `--embed` to use the API.

---
Open to refine the solution, and curious to know how other libraries are using `go-client` in the wild 😄

Reply:
> go-client users need to know about nvim flags and RPC concepts in general. An explicit error fixes 90% of the issue, right?

I agree that explicit error (Method 1) will make the developer-error obvious. It is also an improvement. That's why it was considered as a viable solution together with Method 3.

Consider this: there are many `nvim` flags to consider, and Method 3 makes `go-client` easier to use by setting a sensible value `--embed` as the default. For example, if we know that `--embed` is set for 90% of the use-cases, then we've made the library easier to use for most of our go-client users.

Note that this change doesn't affect the go-client authors who know/remembered about the existence of `--embed`, since they would have instantiated the child process with `ChildProcessArguments("--embed", ...)` in their applications already.

Reply:
> For example, if we know that `--embed` is set for 90% of the use-cases, then we've made the library easier to use for most of our go-client users.

True. But this is more a concern for Nvim, not its clients. 

Also in favor of "fatal error": If we make it an error now, we can always relax that in the future. But the inverse is not true.

Reply:
Mmm, I suppose so. In that case, I'm ok to go with either two methods then.

One final concern: Do you think there are `go-client` users who don't need `--embed`? such that if we go with Method 1 now, then performing log fatal will break their libraries?

Reply:
I'll comment about my opinions for this thread later (Now JST is 2:00 AM :P)